src/Serializer/JsonApiSerializer.php

```diff
@@ -358,7 +358,7 @@ protected function parseRelationships($includedData)
      */
     protected function getIdFromData(array $data)
     {
-        if (!array_key_exists('id', $data)) {
+        if (!array_key_exists('id', $data) || (!is_string($data['id']) && !is_numeric($data['id']))) {
```

Comment:
Makes sense to me. I've updated the PR. Unfortunately that condition will not win any beauty pageants... I've added more test cases as well.